File: packages/vscode-extension/src/providers/base/hierarchy.ts
/**
 * Provider Hierarchy Definitions
 * 
 * Defines the hierarchical structure of objects within a provider.
 * Different providers have different hierarchies:
 * - Unity Catalog: Catalog → Schema → Table/Volume/Function
 * - Hive Metastore: Database → Table
 * - PostgreSQL: Database → Schema → Table/View/Function
 */

export interface HierarchyLevel {
  /** Internal name (e.g., "catalog", "schema", "table") */
  name: string;
  
  /** Display name for UI (e.g., "Catalog", "Schema", "Table") */
  displayName: string;
  
  /** Plural form for UI (e.g., "catalogs", "schemas", "tables") */
  pluralName: string;
  
  /** Optional icon identifier for UI */
  icon?: string;
  
  /** Whether this level supports being a container for other levels */
  isContainer: boolean;
}

export interface ProviderHierarchy {
  /** Ordered list of hierarchy levels from top to bottom */
  levels: HierarchyLevel[];
  
  /**
   * Get hierarchy level at a specific depth (0-indexed)
   * @param depth - Depth in hierarchy (0 = root level)
   */
  getLevel(depth: number): HierarchyLevel | undefined;
  
  /**
   * Get the total depth of the hierarchy
   */
  getDepth(): number;
  
  /**
   * Get level by name
   * @param name - Internal name of the level
   */
  getLevelByName(name: string): HierarchyLevel | undefined;
  
  /**
   * Get the depth of a specific level by name
   * @param name - Internal name of the level
   */
  getLevelDepth(name: string): number;
}

/**
 * Implementation of ProviderHierarchy
 */
export class Hierarchy implements ProviderHierarchy {
  constructor(public readonly levels: HierarchyLevel[]) {
    if (levels.length === 0) {
      throw new Error('Hierarchy must have at least one level');
    }
  }
  
  getLevel(depth: number): HierarchyLevel | undefined {
    return this.levels[depth];
  }
  
  getDepth(): number {
    return this.levels.length;
  }
  
  getLevelByName(name: string): HierarchyLevel | undefined {
    return this.levels.find(level => level.name === name);
  }
  
  getLevelDepth(name: string): number {
    return this.levels.findIndex(level => level.name === name);
  }
}

File: packages/python-sdk/src/schematic/providers/base/hierarchy.py
"""
Provider Hierarchy Definitions

Defines the hierarchical structure of objects within a provider.
"""


from pydantic import BaseModel


class HierarchyLevel(BaseModel):
    """Definition of a single level in the provider hierarchy"""

    name: str  # Internal name (e.g., "catalog", "schema", "table")
    display_name: str  # Display name for UI (e.g., "Catalog")
    plural_name: str  # Plural form for UI (e.g., "catalogs")
    icon: str | None = None  # Optional icon identifier
    is_container: bool  # Whether this level contains other levels


class Hierarchy:
    """Provider hierarchy implementation"""

    def __init__(self, levels: list[HierarchyLevel]):
        if not levels:
            raise ValueError("Hierarchy must have at least one level")
        self.levels = levels

    def get_level(self, depth: int) -> HierarchyLevel | None:
        """Get hierarchy level at a specific depth (0-indexed)"""
        if 0 <= depth < len(self.levels):
            return self.levels[depth]
        return None

    def get_depth(self) -> int:
        """Get the total depth of the hierarchy"""
        return len(self.levels)

    def get_level_by_name(self, name: str) -> HierarchyLevel | None:
        """Get level by internal name"""
        for level in self.levels:
            if level.name == name:
                return level
        return None

    def get_level_depth(self, name: str) -> int:
        """Get the depth of a specific level by name"""
        for i, level in enumerate(self.levels):
            if level.name == name:
                return i
        return -1
